ring/partition_ring.go

```diff
@@ -75,11 +85,227 @@ func (r *PartitionRing) ActivePartitionForKey(key uint32) (int32, error) {
+	if cached := r.shuffleShardCache.getSubringWithLookback(identifier, size, lookbackPeriod, now); cached != nil {
+		return cached, nil
+	}
+
+	subring, err := r.shuffleShard(identifier, size, lookbackPeriod, now)
+	if err != nil {
+		return nil, err
+	}
+
+	r.shuffleShardCache.setSubringWithLookback(identifier, size, lookbackPeriod, now, subring)
+	return subring, nil
+}
+
+func (r *PartitionRing) shuffleShard(identifier string, size int, lookbackPeriod time.Duration, now time.Time) (*PartitionRing, error) {
+	// If the size is too small or too large, run with a size equal to the total number of partitions.
+	// We have to run the function anyway because the logic may filter out some INACTIVE partitions.
+	if size <= 0 || size >= len(r.desc.Partitions) {
+		size = len(r.desc.Partitions)
+	}
+
+	lookbackUntil := now.Add(-lookbackPeriod).Unix()
+
+	// Initialise the random generator used to select instances in the ring.
+	// There are no zones
+	random := rand.New(rand.NewSource(shardUtil.ShuffleShardSeed(identifier, "")))
+
+	// To select one more instance while guaranteeing the "consistency" property,
+	// we do pick a random value from the generator and resolve uniqueness collisions
+	// (if any) continuing walking the ring.
+	tokensCount := len(r.ringTokens)
+
+	result := make(map[int32]struct{}, size)
+	exclude := map[int32]struct{}{}
+
+	for len(result) < size {
+		start := searchToken(r.ringTokens, random.Uint32())
+		iterations := 0
+		found := false
+
+		for p := start; !found && iterations < tokensCount; p++ {
+			iterations++
+
+			// Wrap p around in the ring.
+			if p >= tokensCount {
+				p %= tokensCount
+			}
+
+			pid, ok := r.partitionByToken[Token(r.ringTokens[p])]
+			if !ok {
+				return nil, ErrInconsistentTokensInfo
+			}
+
+			// Ensure the partition has not already been included or excluded.
+			if _, ok := result[pid]; ok {
+				continue
+			}
+			if _, ok := exclude[pid]; ok {
+				continue
+			}
+
+			p, ok := r.desc.Partitions[pid]
+			if !ok {
+				return nil, ErrInconsistentTokensInfo
+			}
+
+			var (
+				shouldInclude bool
+				shouldExtend  bool
+			)
+
+			// Check whether the partition should be included in the result and/or the result set extended.
+			if lookbackPeriod > 0 {
+				if p.IsActive() {
+					// The partition is active. We should include it. The result set is extended only
+					// if the partition became active within the lookback window.
+					shouldInclude = true
+					shouldExtend = p.GetStateTimestamp() >= lookbackUntil
+				} else {
+					// The partition is inactive. We should include it (and then extend the result set)
+					// only if became inactive within the lookback window.
+					shouldInclude = p.GetStateTimestamp() >= lookbackUntil
+					shouldExtend = shouldInclude
+				}
+			} else {
+				// No lookback was provided. We only include active partitions and don't lookback.
+				shouldInclude = p.IsActive()
+				shouldExtend = false
+			}
```

Comment:
I find this slightly easier to read (don't think comments are needed):

```suggestion
			shouldExtend := lookbackPeriod > 0 && p.GetStateTimestamp() >= lookbackUntil
			shouldInclude := p.IsActive() || shouldExtend
```

Reply:
Do you feel strong about this? I find your edit quite cryptic. This logic is complex and I don't think the goal should be compress it as much as possible, but make each case intention as clearest as possible.

Reply:
Discussed offline. Agreed on introducing a 3rd variable to make it easier to understand the rationale: https://github.com/grafana/dskit/pull/476/commits/7552bd65f97657ce2b96e65c37490c0a53484eaa